### DISCUSSION TITLE
Performance implications of a large number of handlers

### DISCUSSION THREAD

**Discussion by sjly3k** [mswjs/msw#2261] [Q&A]
Is there a significant performance difference between declaring only the necessary APIs and pre-declaring many APIs for convenience when testing?

**Comment by kettanaito** [comment]
Hi, @sjly3k. 

MSW performs the request handler lookup by iterating through the entire handlers list stored in memory. Since some handlers can be async and decide how to handle (or not handler) a request within the resolver, the iteration is sequential. Respectively, the more handlers you have, the potentially slower the request lookup gets. 

You never want to load all your handlers in every test. You gain nothing by that. In fact, you are complicating your testing setup and the debugging experience, if something goes wrong. 

What you should do is this:

- Put the happy path handlers in `handlers.ts`, provide them to your `setupServer`/`setupWorker` instance in the setup.
- Use the `.use()` API to issue runtime request handlers on a per-test basis. Don't forget to clean up with `.resetHandlers()` in `afterEach()`.

> [!IMPORTANT]
> Note that this doesn't affect how you _organize_ your handler modules. You can certainly split them on a domain or service basis, improving the maintenance. 

I recommend you read the following:

- [Structuring handlers](https://mswjs.io/docs/best-practices/structuring-handlers)
- [Network behavior overrides](https://mswjs.io/docs/best-practices/network-behavior-overrides)
